### PR TITLE
WIP: Define complex type promotions and support complex binary operators

### PR DESCRIPTION
**Pull request by Roger-luo**
Promotion inherits `numpy.promote_types`'s rule (mostly), but `numpy` do not define promotion between `char` and `complex`. Should we delete promotion for these types too?

Binary operators are not working on tensors and complex values. Still figuring out why...

cc: @ezyang

### CODE REVIEW

File: aten/src/ATen/core/ScalarType.h [hunk 1]
```diff
@@ -165,25 +165,29 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
...
+  constexpr auto c4 = ScalarType::ComplexFloat;
+  constexpr auto c8 = ScalarType::ComplexDouble;
   constexpr auto ud = ScalarType::Undefined;
   if (a == ud || b == ud) {
     return ScalarType::Undefined;
   }
-  if (isComplexType(a) || isComplexType(b)) {
-    AT_ERROR("promoteTypes with complex numbers is not handled yet; figure out what the correct rules should be");
-  }
+  // if (isComplexType(a) || isComplexType(b)) {
```
Comment (ezyang) [inline aten/src/ATen/core/ScalarType.h]: Just delete this now right

File: aten/src/ATen/core/ScalarType.h [hunk 2]
```diff
@@ -165,25 +165,29 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   constexpr auto f2 = ScalarType::Half;
   constexpr auto f4 = ScalarType::Float;
   constexpr auto f8 = ScalarType::Double;
+  constexpr auto c4 = ScalarType::ComplexFloat;
```
Comment (ezyang) [inline aten/src/ATen/core/ScalarType.h]: nit: Can you name this c8 instead? That'll be consistent with `torch.complex64` and `torch.complex128`. c16 is too big, but you could maybe try `cd` instead...

**Comment by ezyang** [comment]
It looks like our vectorizer doesn't know how to deal with complex numbers, unfortunately. Happy to take the promotion logic patch separately.

File: aten/src/ATen/core/ScalarType.h (hunk 2 above)
Comment (Roger-luo) [reply aten/src/ATen/core/ScalarType.h]: I thought `8` means the precision, so we have `c4` for float precision complex and `c8` for double precision complex which is consistent to `f4` and `f8`...

Comment (ezyang) [reply aten/src/ATen/core/ScalarType.h]: oh all right :)

**Comment by Roger-luo** [comment]
Discussed on slack. We will allow promotion between `char` and `complex` for now (this is not the same with numpy).

File: torch/csrc/utils/python_arg_parser.cpp [hunk 3]
```diff
@@ -222,6 +225,8 @@ void FunctionParameter::set_default_str(const std::string& str) {
     default_bool = (str == "True" || str == "true");
   } else if (type_ == ParameterType::DOUBLE) {
     default_double = atof(str.c_str());
+  } else if (type_ == ParameterType::COMPLEX) {
+    default_complex[0] = atof(str.c_str()); // TODO: parse "x + xj"?
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp]: Don't you have to explicitly initialize `default_complex[1]` here too? Not that any of this is exercised at the moment.

**Comment by SsnL** [comment]
@Roger-luo This needs a rebase to trigger CI again. Also, I wonder if we can some tests for this.

**Comment by Roger-luo** [comment]
@SsnL thanks!

I'll need to fix the CI error I got with `AT_DISPATCH_ALL_TYPES_AND_COMPLEX` first... It works find on my Mac but got error on Ubuntu :-(

Then I'll be able to add some tests.

**Comment by ezyang** [comment]
(For the record: I helped Roger repro the issue offline)

**Comment by PhilSee** [comment]
On the `AT_DISPATCH_ALL_TYPES ` issue: This might actually be related to a bug in GCC, where the visibility of a lambda and its captured objects list is treated differently by GCC: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80947 . Adding a `__attribute__ ((visibility("hidden")))` to the tensor_cpu declaration in TensorFactories.cpp:643 is a suggested workaround that actually does the trick for me. It has some pitfalls, however. More on Slack.

**Comment by Roger-luo** [comment]
Okay... this is related to different gcc versions, MWE:

```c++
#include <iostream>
#include <complex>

template <class T>
struct Vec256 {
public:
 T values[32 / sizeof(T)] = {0};
  static constexpr int size = 32 / sizeof(T);
  Vec256() {}
  Vec256(T val) {
    for (int i = 0; i != size; i++) {
      values[i] = val;
    }
  }
};

int main() {
  Vec256<std::complex<double>> r(1);
  std::cout << r.values[1] << std::endl;
  return 0;
}
```

This works fine with `clang` and later version of `g++`, but won't built with `g++` 4.8. specialize the `class` manually for complex will work.

File: aten/src/ATen/cpu/vec256/vec256.h [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
 
+#include <complex>
 #include "intrinsics.h"
 
 #include "vec256_base.h"
 #include "vec256_float.h"
 #include "vec256_double.h"
+#include "vec256_complex.h"
+// #include "vec256_complex_float.h"
+// #include "vec256_complex_double.h"
```
Comment (ezyang) [inline aten/src/ATen/cpu/vec256/vec256.h]: Don't forget to delete these comments when you're done.

Comment (Roger-luo) [reply aten/src/ATen/cpu/vec256/vec256.h]: That's SIMDs for complex number, I'll uncomment it later.

**Comment by Roger-luo** [comment]
It seems that something breaks cuda tests:

```
22:56:08 ERROR: test_mul_intertype_scalar (__main__.TestCuda)
22:56:08 ----------------------------------------------------------------------
22:56:08 Traceback (most recent call last):
22:56:08   File "/var/lib/jenkins/workspace/test/common.py", line 275, in wrapper
22:56:08     method(*args, **kwargs)
22:56:08   File "test_cuda.py", line 925, in test_mul_intertype_scalar
22:56:08     test_mul(torch.float16)
22:56:08   File "test_cuda.py", line 922, in test_mul
22:56:08     x *= y
22:56:08 RuntimeError: TensorIterator expected type torch.cuda.DoubleTensor but got torch.cuda.HalfTensor[]
```

Is it a type promotion problem?

**Comment by ezyang** [comment]
Looks like a promotion problem to me!

By the way, @gchanan brought to my attention that we have a bunch of latent bugs in PyTorch, owing to the fact that in a number of places we do `if (scalar.is_float()) { /* assume it's float */ } else { /* assume it's int */ }`. We'll need to fix these. Maybe this is causing the error you see above.

**Comment by ezyang** [comment]
@Roger-luo Do you need any help with this patch?

**Comment by Roger-luo** [comment]
Ah I think I should just close this PR since there's complex number now!

**Pull request closed without merging**